src/header.rs

```diff
@@ -0,0 +1,107 @@
+//! headers are sets of consecutive keywords and tokens, such as
+//! `pub const unsafe fn foo` and `pub(crate) unsafe trait Bar`.
+//!
+//! This module contains general logic for formatting such headers,
+//! where they are always placed on a single line except when there
+//! are comments between parts of the header.
+
+use std::borrow::Cow;
+
+use rustc_ast as ast;
+use rustc_span::symbol::Ident;
+use rustc_span::{Span, DUMMY_SP};
+
+use crate::comment::combine_strs_with_missing_comments;
+use crate::rewrite::RewriteContext;
+use crate::shape::Shape;
+use crate::utils::rewrite_ident;
+
+pub(crate) fn format_header(
+    cx: &RewriteContext<'_>,
+    shape: Shape,
+    parts: Vec<HeaderPart>,
+) -> String {
+    debug!(?parts, "format_header");
+    let shape = shape.infinite_width();
+    let mut parts = parts.into_iter().filter(|x| !x.snippet.is_empty());
+    let Some(part) = parts.next() else {
+        return String::new();
+    };
+
+    let mut st = part.snippet.into_owned();
+    let mut span = part.span;
+
+    for part in parts {
+        debug!(?st, "before combine");
+        st = combine_strs_with_missing_comments(
+            cx,
+            &st,
+            &part.snippet,
+            span.between(part.span),
+            shape,
+            true,
+        )
+        .unwrap_or_else(|| format!("{} {}", &st, part.snippet));
+        debug!(?st);
+        span = part.span;
+    }
+
+    st
+}
+
+#[derive(Debug)]
+pub(crate) struct HeaderPart {
+    /// snippet of this part without surrounding space
+    snippet: Cow<'static, str>,
+    span: Span,
+}
+
+impl HeaderPart {
+    pub(crate) fn new(snippet: impl Into<Cow<'static, str>>, span: Span) -> Self {
+        Self {
+            snippet: snippet.into(),
+            span,
+        }
+    }
+
+    /// creates an empty header part. This is ignored by `format_header`.
+    fn empty() -> Self {
+        Self {
+            snippet: "".into(),
+            span: DUMMY_SP,
+        }
+    }
+
+    pub(crate) fn ident(cx: &RewriteContext<'_>, ident: Ident) -> Self {
+        Self {
+            snippet: cx.snippet(ident.span).to_owned().into(),
+            span: ident.span,
+        }
+    }
+
+    pub(crate) fn visibility(cx: &RewriteContext<'_>, vis: &ast::Visibility) -> Self {
+        let snippet = match vis.kind {
+            ast::VisibilityKind::Public => Cow::from("pub"),
+            ast::VisibilityKind::Inherited => return Self::empty(),
```

Comment:
Any downside to using `Cow::from("")` here instead of early returning `Self::empty()` so that the `HeaderPart` has the same `span` as the `ast::Visibility`? Either way, the logic in format_header should ignore this.